examples/server/server.cpp

```diff
@@ -163,6 +164,10 @@ struct server_slot {
     // when a task is submitted, we first tokenize the prompt and store it here
     std::vector<llama_token> prompt_tokens;
 
+    llama_ngram_cache nc_context;
+    std::vector<llama_token> draft;
+    std::vector<llama_token> context_tokens;
```

Comment:
Isn't this the same as `cache_tokens`?

Ideally, the `llama_sampling_context` should maintain a history of the processed tokens. There are already steps in that direction via the `std::vector<llama_token> prev;` member and the `llama_sampling_accept()` API, but some more work would be needed (such as API for removing discarded tokens). Not needed to be done in this PR - just clarifying the long-term goals

Reply:
Yes, I think it should be possible to refactor the code in such a way that the same vector is used for caching and lookup.